gems/sorbet-runtime/lib/types/types/attached_class.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+# typed: true
+
+module T::Types
+  # Modeling AttachedClass properly at runtime would require additional
+  # tracking, so at runtime we permit all values and rely on the static checker.
+  class AttachedClassType < Base
+
+    def initialize(); end
+
+    # @override Base
+    def name
+      "AttachedClass"
+    end
+
+    # @override Base
+    def valid?(obj)
+      true
+    end
+
+    # @override Base
+    private def subtype_of_single?(other)
+      case other
+      when AttachedClassType
+          true
+        else
+          false
+      end
+    end
+  end
+end
+
+AttachedClass = T.unsafe(T::Types::AttachedClassType.new)
```

Comment:
I wonder if it's worth making this into `T::AttachedClass`, sort of how `T::Array` works specially for the purpose of type syntax but internally becomes the same `Array` symbol. It would be kind of nice to not have to pollute the global namespace.

Reply:
It's also kind of strange this way because the static system and runtime system actually resolve to two different symbols:

- the static system resolves `AttachedClass` in class `A` to `A::AttachedClass`
- the runtime system resolves `AttachedClass` in `A` to `::AttachedClass`

which again, works out in practice, it's just kind of odd, and might be less odd of it was explicitly called out with dedicated syntax like `T::AttachedClass`. Because while under the hood it's implemented via a type_template, to a user this is just a feature built into Sorbet so it could justifiably have special syntax.

Reply:
I like putting it in the `T` namespace better, but it does raise the question of how you would refer to the `AttachedClass` of another class. With the implementation as it is, you can say `IO::AttachedClass`, but moving it into the `T` namespace would prohibit that.

Reply:
I think I'm going to punt on this decision, and address it in a follow up PR.